src/queue.cpp

```diff
@@ -561,6 +540,87 @@ cl_int cvk_command_kernel::dispatch_uniform_region(
     return CL_SUCCESS;
 }
 
+cl_int cvk_command_kernel::dispatch_uniform_region(
+    const cvk_ndrange& region, cvk_command_buffer& command_buffer) {
+
+    // Calculate number of workgroups for region
+    std::array<uint32_t, 3> num_workgroups;
+    for (cl_uint i = 0; i < 3; i++) {
+        CVK_ASSERT(region.gws[i] % region.lws[i] == 0);
+        num_workgroups[i] = region.gws[i] / region.lws[i];
+    };
+
+    auto program = m_kernel->program();
+    auto& vklimits = m_queue->device()->vulkan_limits();
+    if (!program->push_constant(pushconstant::region_offset) ||
+        !program->push_constant(pushconstant::region_group_offset)) {
+        for (cl_uint i = 0; i < 3; ++i) {
+            if (num_workgroups[i] > vklimits.maxComputeWorkGroupCount[i]) {
+                cvk_error_fn(
+                    "global work size (%d, %d, %d) exceeds device limits"
+                    " of (%d, %d, %d)",
+                    num_workgroups[0], num_workgroups[1], num_workgroups[2],
+                    vklimits.maxComputeWorkGroupCount[0],
+                    vklimits.maxComputeWorkGroupCount[1],
+                    vklimits.maxComputeWorkGroupCount[2]);
+                cvk_error_fn(
+                    "Compiling the kernel with either '-cl-std=2.0', "
+                    "'-cl-std=3.0' or 'cl-arm-non-uniform-work-group-size' "
+                    "might allow to exceed the device limits");
+
+                return CL_INVALID_WORK_ITEM_SIZE;
+            }
+        }
+    }
+
+    if (region.lws[0] * region.lws[1] * region.lws[2] >
+        vklimits.maxComputeWorkGroupInvocations) {
+        return CL_INVALID_WORK_GROUP_SIZE;
+    }
+
+    for (int i = 0; i < 3; i++) {
+        if (region.lws[i] > vklimits.maxComputeWorkGroupSize[i]) {
+            return CL_INVALID_WORK_ITEM_SIZE;
+        }
+    }
+
+#define FOR_REGION_WITHIN_LIMITS(dim, nb_wg, vklim, reg, reg_off, reg_gws)     \
```

Comment:
I'm not sure to get what you would like exactly.
I got rid of the macro, and use a recursive function instead.
I hope it will be easier to read like that.